File: frontend/src/api/readerBackground.ts
import http from './http'

export type ReaderBackgroundFit = 'cover' | 'contain'
export type ReaderBackgroundPosition = 'top' | 'center' | 'bottom'

export interface ReaderBackgroundSettings {
  enabled: boolean
  readerEnabled?: boolean
  fit: ReaderBackgroundFit
  position: ReaderBackgroundPosition
  overlay: number
}

export interface ReaderBackgroundMetadata extends ReaderBackgroundSettings {
  contentType: string
  byteSize: number
  updatedAt: number
  etag: string
}

export interface ReaderBackgroundImage {
  blob: Blob
  etag: string
}

export function fetchReaderBackgroundMetadata() {
  return http
    .get<ReaderBackgroundMetadata | null>('/reader-background')
    .then((response) => response.data)
}

export function fetchReaderBackgroundImage() {
  return http
    .get<Blob>('/reader-background/image', { responseType: 'blob' })
    .then((response) => ({
      blob: response.data,
      etag: String(response.headers.etag || '').replace(/^W\//, '').replace(/^"|"$/g, ''),
    }))
}

export function uploadReaderBackground(image: Blob, settings: ReaderBackgroundSettings) {
  const form = new FormData()
  form.append('file', image, backgroundFileName(image.type))
  form.append('enabled', String(settings.enabled))
  form.append('readerEnabled', String(settings.readerEnabled ?? false))
  form.append('fit', settings.fit)
  form.append('position', settings.position)
  form.append('overlay', String(settings.overlay))
  return http
    .post<ReaderBackgroundMetadata>('/reader-background', form, {
      headers: { 'Content-Type': 'multipart/form-data' },
    })
    .then((response) => response.data)
}

export function updateReaderBackgroundSettings(settings: ReaderBackgroundSettings) {
  return http
    .post<ReaderBackgroundMetadata>('/reader-background/settings', settings)
    .then((response) => response.data)
}

export function removeReaderBackground() {
  return http
    .delete<{ deleted: boolean }>('/reader-background')
    .then((response) => response.data)
}

function backgroundFileName(contentType: string) {
  const extension = contentType === 'image/jpeg'
    ? 'jpg'
    : contentType.replace(/^image\//, '') || 'bin'
  return `reader-background.${extension}`
}

File: src/model/reader_background.rs
use crate::error::error::AppError;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ReaderBackgroundFit {
    Cover,
    Contain,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ReaderBackgroundPosition {
    Top,
    Center,
    Bottom,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReaderBackgroundSettings {
    pub enabled: bool,
    #[serde(default)]
    pub reader_enabled: bool,
    pub fit: ReaderBackgroundFit,
    pub position: ReaderBackgroundPosition,
    pub overlay: f32,
}

impl Default for ReaderBackgroundSettings {
    fn default() -> Self {
        Self {
            enabled: true,
            reader_enabled: false,
            fit: ReaderBackgroundFit::Cover,
            position: ReaderBackgroundPosition::Center,
            overlay: 0.45,
        }
    }
}

impl ReaderBackgroundSettings {
    pub fn validate(&self) -> Result<(), AppError> {
        if !self.overlay.is_finite() || !(0.0..=0.9).contains(&self.overlay) {
            return Err(AppError::BadRequest(
                "背景遮罩强度必须在 0 到 0.9 之间".to_string(),
            ));
        }
        Ok(())
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReaderBackgroundMetadata {
    pub content_type: String,
    pub byte_size: u64,
    pub updated_at: i64,
    pub etag: String,
    #[serde(flatten)]
    pub settings: ReaderBackgroundSettings,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn settings_reject_invalid_overlay() {
        for overlay in [-0.01, 0.91, f32::NAN, f32::INFINITY] {
            let settings = ReaderBackgroundSettings {
                overlay,
                ..ReaderBackgroundSettings::default()
            };
            assert!(settings.validate().is_err());
        }
    }

    #[test]
    fn legacy_settings_keep_reader_background_disabled() {
        let settings: ReaderBackgroundSettings = serde_json::from_str(
            r#"{"enabled":true,"fit":"cover","position":"center","overlay":0.45}"#,
        )
        .unwrap();

        assert!(settings.enabled);
        assert!(!settings.reader_enabled);
    }
}
